lighthouse-core/report/html/components/gauge.js

```diff
@@ -0,0 +1,398 @@
+  top: calc(var(--score-container-padding) + var(--gauge-circle-size) / 2);
+}
+
+.lh-category .lh-gauge__percentage {
+  --gauge-circle-size: var(--gauge-circle-size-big);
+  --gauge-percentage-font-size: var(--gauge-percentage-font-size-big);
+}
+
+.lh-gauge__wrapper {
+  position: relative;
+  display: flex;
+  align-items: center;
+  flex-direction: column;
+  text-decoration: none;
+  padding: var(--score-container-padding);
+
+  --circle-border-width: 8;
+  --transition-length: 1s;
+
+  /* Contain the layout style paint & layers during animation*/
+  contain: content;
+  will-change: opacity; /* Only using for layer promotion */
+}
+
+.lh-gauge__label {
+  font-size: var(--gauge-label-font-size);
+  line-height: var(--gauge-label-line-height);
+  margin-top: 10px;
+  text-align: center;
+  color: var(--report-text-color);
+}
+
+/* TODO(#8185) use more BEM (.lh-gauge__label--big) instead of relying on descendant selector */
+.lh-category .lh-gauge__label {
+  --gauge-label-font-size: var(--gauge-label-font-size-big);
+  --gauge-label-line-height: var(--gauge-label-line-height-big);
+  margin-top: 14px;
+}
+  </style>
+
+  <a href="#" class="lh-gauge__wrapper">
+    <!-- Wrapper exists for the ::before plugin icon. Cannot create pseudo-elements on svgs. -->
+    <div class="lh-gauge__svg-wrapper">
+      <svg viewBox="0 0 120 120" class="lh-gauge">
+        <circle class="lh-gauge-base" r="56" cx="60" cy="60"></circle>
+        <circle class="lh-gauge-arc" transform="rotate(-90 60 60)" r="56" cx="60" cy="60"></circle>
+      </svg>
+    </div>
+    <div class="lh-gauge__percentage"></div>
+    <div class="lh-gauge__label"></div>
+  </a>
+`;
+
+const attributes = {
+  score: '0.5',
+  variant: 'pass',
+  label: 'Lighthouse',
+};
+
+class LighthouseGaugeElement extends HTMLElement {
+  constructor() {
+    super();
+    this._shadowRoot = this.attachShadow({mode: 'open'});
+    this._shadowRoot.appendChild(template.content.cloneNode(true));
+    this._attributes = {...attributes};
+
+    for (const key in Object.keys(attributes)) {
+      const attributeName = /** @type {keyof typeof attributes} */ (key);
+      Object.defineProperty(this, attributeName, {
+        get: () => this._attributes[attributeName],
+        set: value => this.setAttribute(attributeName, String(value)),
+      });
+    }
+
+    this._render();
+  }
+
+  static get observedAttributes() {
+    return Object.keys(attributes);
+  }
+
+
+  connectedCallback() {
+    for (const attribute of this.attributes) {
+      // @ts-ignore
+      this[attribute.name] = attribute.nodeValue;
+    }
+  }
+
+  /** @param {string} name @param {string} oldValue @param {string} newValue */
+  attributeChangedCallback(name, oldValue, newValue) {
+    if (!(name in attributes) || oldValue === newValue) return;
+
+    // @ts-ignore
+    this._attributes[name] = newValue;
+    this._render();
+  }
+
+  _render() {
+    const wrapper = this._shadowRoot.querySelector('.lh-gauge__wrapper');
```

Comment:
I think this might come down to if we're able to share code between our components, if we can then 10K for lit-html might be really nice, but if we can't, paying an extra 10K per component is too steep. How is this typically handled? in-browser modules?

Reply:
It's ~3.3k gzipped :)

And yes, usually a single module instance is shared by all components.